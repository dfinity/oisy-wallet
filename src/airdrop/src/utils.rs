use crate::CanisterError::{UnknownOisyWalletAddress, CanisterKilled, GeneralError, NoMoreCodes};
use crate::{CustomResult, EthereumAddress, Code, AirdropAmount};
use candid::Principal;
use ic_cdk::api::call::CallResult;
use ic_cdk::{call, caller};

use crate::{KILLED, TOTAL_TOKENS, PRINCIPALS_USER_ETH, AIRDROP_REWARD, PRE_GENERATED_CODES};


// TODO: add to initial parameters
static BACKEND: &str = "a3shf-5eaaa-aaaaa-qaafa-cai";

pub async fn get_eth_address() -> CustomResult<EthereumAddress> {
    let args = caller();

    let result: CallResult<(String,)> = call(
        Principal::from_text(BACKEND).unwrap(),
        "eth_address_of",
        (args,),
    )
    .await;

    match result {
        Err((_, _message)) => Err(UnknownOisyWalletAddress),
        Ok((eth_address,)) => Ok(EthereumAddress(eth_address)),
    }
}

/// Helper function to check if the cannister is still alive
pub fn check_if_killed() -> CustomResult<()> {
    KILLED.with(|killed| {
        let killed = killed.borrow();
        if *killed {
            Err(CanisterKilled)
        } else {
            Ok(())
        }
    })
}

/// Deduct token
pub fn deduct_tokens(amount: u64) -> CustomResult<()> {
    TOTAL_TOKENS.with(|total_tokens| {
        let mut total_tokens = total_tokens.borrow_mut();
        if *total_tokens < amount {
            Err(GeneralError(
                "Not enough tokens left".to_string(),
            ))
        } else {
            *total_tokens -= amount;
            Ok(())
        }
    })
}

/// Register a principal with an ethereum address
pub fn register_principal_with_eth_address(
    principal: Principal,
    code: Code,
    eth_address: EthereumAddress,
) {
    PRINCIPALS_USER_ETH.with(|principal_eth| {
        let mut principal_eth = principal_eth.borrow_mut();
        principal_eth.insert(principal, (code, eth_address));
    });
}

/// Add user to the list of users to send tokens to
pub fn add_user_to_airdrop_reward(eth_address: EthereumAddress, amount: AirdropAmount) {
    AIRDROP_REWARD.with(|airdrop_reward| {
        let mut airdrop_reward = airdrop_reward.borrow_mut();
        airdrop_reward.push((eth_address, amount));
    });
}

// "generate" codes
pub fn get_pre_codes() -> CustomResult<Code> {
    PRE_GENERATED_CODES.with(|pre_generated_codes| {
        let mut pre_generated_codes = pre_generated_codes.borrow_mut();
        match pre_generated_codes.pop() {
            Some(code) => Ok(code),
            None => Err(NoMoreCodes),
        }
    })
}

