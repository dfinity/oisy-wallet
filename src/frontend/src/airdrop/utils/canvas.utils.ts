export const canvasToBlob = async ({
	canvas,
	type
}: {
	canvas: HTMLCanvasElement;
	type: string;
}): Promise<Blob | null> => {
	return new Promise<Blob | null>((resolve) =>
		canvas.toBlob((blob: Blob | null) => resolve(blob), type)
	);
};
