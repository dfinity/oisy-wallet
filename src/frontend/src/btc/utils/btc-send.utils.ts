import { BTC_DECIMALS } from '$env/tokens.btc.env';

export const convertNumberToSatoshis = ({ amount }: { amount: number | string }): bigint =>
	BigInt(Number(amount) * 10 ** BTC_DECIMALS);
