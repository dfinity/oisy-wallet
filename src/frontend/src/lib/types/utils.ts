export type RequiredExcept<T, K extends keyof T> = Required<Omit<T, K>> & Pick<T, K>;

export type ResultSuccess<T = unknown> = { success: boolean; err?: T };

export type ResultSuccessReduced<T = unknown> = ResultSuccess<T[]>;

// We disable the eslint rule here because this is the utility type that we use to define such rule.
// eslint-disable-next-line local-rules/use-option-type-wrapper
export type Option<T> = T | null | undefined;
