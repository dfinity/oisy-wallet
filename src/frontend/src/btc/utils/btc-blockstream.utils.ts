import { CONFIRMED_BTC_TRANSACTION_MIN_CONFIRMATIONS } from '$btc/constants/btc.constants';
import type { BtcTransactionUi } from '$btc/types/btc';
import type { BtcAddress } from '$lib/types/address';
import type { BlockstreamTransaction } from '$lib/types/blockstream';
import { nonNullish } from '@dfinity/utils';

export const mapBtcTransaction = ({
	transaction: { vin, vout, txid, status, fee },
	btcAddress,
	latestBitcoinBlockHeight
}: {
	transaction: BlockstreamTransaction;
	btcAddress: BtcAddress;
	latestBitcoinBlockHeight: number;
}): BtcTransactionUi => {
	// Step 1: Calculate total input value and determine if the trabtcAddressTransactionsnsaction is a "send"
	const { totalInputValue, isTypeSend } = vin.reduce<{
		totalInputValue: number;
		isTypeSend: boolean;
	}>(
		(acc, { prevout }) => {
			// For coinbase transactions, prevout might not exist
			if (!prevout) {
				return acc;
			}

			return {
				totalInputValue: acc.totalInputValue + prevout.value,
				isTypeSend: acc.isTypeSend ? acc.isTypeSend : prevout.scriptpubkey_address === btcAddress
			};
		},
		{
			totalInputValue: 0,
			isTypeSend: false
		}
	);

	// Step 2: Analyse outputs to determine total value, destination addresses, and whether it's a self-transaction
	const { totalOutputValue, totalValue, to, selfTransaction } = vout.reduce<{
		to: string[];
		totalValue: number | undefined;
		totalOutputValue: number;
		selfTransaction: boolean;
	}>(
		(acc, output) => {
			const { scriptpubkey_address: addr, value } = output;

			// Skip outputs without addresses (e.g., OP_RETURN outputs)
			if (!addr) {
				return {
					...acc,
					totalOutputValue: acc.totalOutputValue + value,
					selfTransaction: false
				};
			}

			// For 'send' tx: include outputs not to the sender
			// For 'receive' tx: include only outputs to the user
			const isValidOutput =
				(isTypeSend && addr !== btcAddress) || (!isTypeSend && addr === btcAddress);

			return {
				...acc,
				totalValue: isValidOutput ? (acc.totalValue ?? 0) + value : acc.totalValue,
				to: isValidOutput ? [...acc.to, addr] : acc.to,
				totalOutputValue: acc.totalOutputValue + value,
				// If all outputs are to the sender, it's a self-transaction
				selfTransaction: acc.selfTransaction && addr === btcAddress
			};
		},
		{
			totalOutputValue: 0,
			totalValue: undefined,
			to: [],
			selfTransaction: true
		}
	);

	// Step 3: Calculate the transaction fee (provided directly by Blockstream API)
	const utxosFee = fee;

	// Step 4: Compute the number of confirmations
	// +1 is needed to account for the block where the transaction was first included
	const confirmations = nonNullish(status.block_height)
		? latestBitcoinBlockHeight - status.block_height + 1
		: undefined;

	// Step 5: Derive transaction status based on confirmation status
	const transactionStatus = !status.confirmed
		? 'pending'
		: nonNullish(confirmations) && confirmations >= CONFIRMED_BTC_TRANSACTION_MIN_CONFIRMATIONS
			? 'confirmed'
			: 'unconfirmed';

	// Step 6: Determine the value of the transaction
	const value = selfTransaction
		? // For self-transactions, we consider the total output value plus the fee, since that is logically the value spent for the presumable 'send' transaction
			BigInt(totalOutputValue + utxosFee)
		: nonNullish(totalValue)
			? // For 'send' transactions, we consider the total value sent plus the fee, since that is logically the value spent for the transaction
				BigInt(isTypeSend ? totalValue + utxosFee : totalValue)
			: undefined;

	// Step 7: Compose the final structured BTC transaction object for the UI
	return {
		id: txid,
		timestamp: nonNullish(status.block_time)
			? BigInt(status.block_time)
			: BigInt(Date.now() / 1000),
		value,
		status: transactionStatus,
		blockNumber: status.block_height ?? undefined,
		type: isTypeSend ? 'send' : 'receive',
		from: isTypeSend ? btcAddress : (vin[0]?.prevout?.scriptpubkey_address ?? 'coinbase'),
		to: selfTransaction ? [btcAddress] : to,
		confirmations
	};
};
