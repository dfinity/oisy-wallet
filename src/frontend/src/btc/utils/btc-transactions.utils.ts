import {
	CONFIRMED_BTC_TRANSACTION_MIN_CONFIRMATIONS,
	PENDING_BTC_TRANSACTION_MIN_CONFIRMATIONS
} from '$btc/constants/btc.constants';
import type { BtcTransactionUi } from '$btc/types/btc';
import type { BtcAddress } from '$lib/types/address';
import type { BitcoinTransaction } from '$lib/types/blockchain';
import { isNullish, nonNullish } from '@dfinity/utils';

// TODO: Add docs/comments to the steps needed to parse a BTC tx
export const mapBtcTransaction = ({
	transaction: { inputs, block_index, out, hash, time },
	btcAddress,
	latestBitcoinBlockHeight
}: {
	transaction: BitcoinTransaction;
	btcAddress: BtcAddress;
	latestBitcoinBlockHeight: number;
}): BtcTransactionUi => {
	const { totalInputValue, isTypeSend } = inputs.reduce<{
		totalInputValue: number;
		isTypeSend: boolean;
	}>(
		(acc, { prev_out: { value, addr } }) => ({
			totalInputValue: acc.totalInputValue + value,
			isTypeSend: acc.isTypeSend ? acc.isTypeSend : addr === btcAddress
		}),
		{
			totalInputValue: 0,
			isTypeSend: false
		}
	);

	const { totalOutputValue, value, to } = out.reduce<{
		totalOutputValue: number;
		value: number | undefined;
		to: string | undefined;
	}>(
		(acc, { addr, value }) => {
			// TODO: test what happens when user sends to the current address (hence isValidOutput = false)
			const isValidOutput =
				(isTypeSend && addr !== btcAddress) || (!isTypeSend && addr === btcAddress);

			if (isNullish(acc.value) && isValidOutput) {
				acc.value = (acc.value ?? 0) + value;
			}

			if (isNullish(acc.to) && isValidOutput) {
				acc.to = addr;
			}

			return {
				...acc,
				totalOutputValue: acc.totalOutputValue + value
			};
		},
		{
			totalOutputValue: 0,
			value: undefined,
			to: undefined
		}
	);

	const utxosFee = totalInputValue - totalOutputValue;

	const confirmations = nonNullish(block_index)
		? latestBitcoinBlockHeight - block_index
		: undefined;

	const status =
		isNullish(confirmations) || confirmations <= PENDING_BTC_TRANSACTION_MIN_CONFIRMATIONS
			? 'pending'
			: confirmations >= CONFIRMED_BTC_TRANSACTION_MIN_CONFIRMATIONS
				? 'confirmed'
				: 'unconfirmed';

	return {
		id: hash,
		timestamp: BigInt(time),
		value: nonNullish(value) ? BigInt(isTypeSend ? value + utxosFee : value) : undefined,
		status,
		blockNumber: block_index ?? undefined,
		type: isTypeSend ? 'send' : 'receive',
		from: isTypeSend ? btcAddress : inputs[0].prev_out.addr,
		to,
		confirmations
	};
};
