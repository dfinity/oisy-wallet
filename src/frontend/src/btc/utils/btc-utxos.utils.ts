import { ZERO } from '$lib/constants/app.constants';
import type { Utxo } from '@dfinity/ckbtc';
import { uint8ArrayToHexString } from '@dfinity/utils';

export interface UtxoSelectionResult {
	selectedUtxos: Utxo[];
	totalInputValue: bigint;
	changeAmount: bigint;
	sufficientFunds: boolean;
	feeSatoshis: bigint;
}

/**
 * Converts a UTXO transaction ID (Uint8Array) to a hex string
 */
export const utxoTxIdToString = (txid: Uint8Array | number[]): string =>
	uint8ArrayToHexString(txid);

/**
 * Extracts transaction IDs from an array of UTXOs
 */
export const extractUtxoTxIds = (utxos: Utxo[]): string[] =>
	utxos.map(({ outpoint: { txid } }) => utxoTxIdToString(txid));

/**
 * Estimates transaction size in bytes based on number of inputs and outputs
 * This is a simplified calculation for P2WPKH transactions
 */
export const estimateTransactionSize = ({
	numInputs,
	numOutputs
}: {
	numInputs: number;
	numOutputs: number;
}): number => {
	// version (4) and locktime (4) and input count (1) and output count (1)
	const baseSize = 10;

	// P2WPKH input size: outpoint (36) and scriptSig length (1) and scriptSig (0) and sequence (4) = 41 bytes
	// Plus witness data: witness stack items (1) and signature (72) and pubkey (33) = 106 bytes
	// But witness data is counted as 1/4 for fee calculation, so effective size is 41 and 106/4 = 67.5 bytes
	const inputSize = 68;

	// P2WPKH output size: value (8) and scriptPubKey length (1) and scriptPubKey (22) = 31 bytes
	const outputSize = 31;

	return baseSize + numInputs * inputSize + numOutputs * outputSize;
};

/**
 * Main function to select UTXOs for a transaction with fee consideration
 * This function iteratively selects UTXOs and calculates fees until sufficient funds are found
 */
export const calculateUtxoSelection = ({
	availableUtxos,
	amountSatoshis,
	feeRateMiliSatoshisPerVByte
}: {
	availableUtxos: Utxo[];
	amountSatoshis: bigint;
	feeRateMiliSatoshisPerVByte: bigint;
}): UtxoSelectionResult => {
	if (availableUtxos.length === 0) {
		return {
			selectedUtxos: [],
			totalInputValue: ZERO,
			changeAmount: ZERO,
			sufficientFunds: false,
			feeSatoshis: ZERO
		};
	}

	// Sort UTXOs by value in descending order
	const sortedUtxos = [...availableUtxos].sort((a, b) => {
		const aValue = BigInt(a.value);
		const bValue = BigInt(b.value);
		return aValue > bValue ? -1 : aValue < bValue ? 1 : 0;
	});

	const selectedUtxos: Utxo[] = [];
	let totalInputValue = ZERO;
	let estimatedFee = ZERO;

	// Iteratively add UTXOs until we have enough to cover amount and fee
	for (const utxo of sortedUtxos) {
		selectedUtxos.push(utxo);
		totalInputValue += BigInt(utxo.value);

		// Calculate current transaction size and fee
		// 2 outputs: one for recipient, one for change (if needed)
		const txSize = estimateTransactionSize({
			numInputs: selectedUtxos.length,
			numOutputs: 2
		});
		estimatedFee = (BigInt(txSize) * feeRateMiliSatoshisPerVByte) / 1000n;
		const totalRequired = amountSatoshis + estimatedFee;

		// Check if we have enough to cover amount and fee
		if (totalInputValue >= totalRequired) {
			const changeAmount = totalInputValue - totalRequired;

			return {
				selectedUtxos,
				totalInputValue,
				changeAmount,
				sufficientFunds: true,
				feeSatoshis: estimatedFee
			};
		}
	}

	// If we get here, we don't have sufficient funds
	return {
		selectedUtxos,
		totalInputValue,
		changeAmount: ZERO,
		sufficientFunds: false,
		feeSatoshis: estimatedFee
	};
};

/**
 * Filters out UTXOs that are locked by pending transactions
 */
export const filterLockedUtxos = ({
	utxos,
	pendingTxIds
}: {
	utxos: Utxo[];
	pendingTxIds: string[];
}): Utxo[] =>
	utxos.filter((utxo) => {
		const txIdHex = utxoTxIdToString(utxo.outpoint.txid);
		return !pendingTxIds.includes(txIdHex);
	});

/**
 * Filters UTXOs based on confirmations and excludes those locked by pending transactions
 */
export const filterAvailableUtxos = ({
	utxos,
	options
}: {
	utxos: Utxo[];
	options: {
		minConfirmations: number;
		pendingTxIds: string[];
	};
}): Utxo[] => {
	const { minConfirmations, pendingTxIds } = options;

	// First filter by confirmations to ensure transaction security
	// Note: UTXOs are pre-filtered by the Bitcoin canister endpoint
	const confirmedUtxos = utxos.filter((utxo) => {
		// Unconfirmed transactions have height 0
		if (utxo.height === 0) {
			return false;
		}

		// Check if it has enough confirmations
		return utxo.height >= minConfirmations;
	});

	// Then filter out locked UTXOs
	return filterLockedUtxos({
		utxos: confirmedUtxos,
		pendingTxIds
	});
};
