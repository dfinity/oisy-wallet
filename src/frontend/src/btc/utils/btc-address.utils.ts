import type { BtcAddress } from '$declarations/backend/backend.did';
import { assertNever } from '$lib/types/utils';
import {
	BtcAddressType,
	parseBtcAddress as parseBtcAddressCkbtc,
	type BtcAddressInfo
} from '@dfinity/ckbtc';

const createBtcAddressFromAddressInfo = ({ info }: { info: BtcAddressInfo }): BtcAddress => {
	switch (info.type) {
		case BtcAddressType.P2wpkhV0:
			return { P2WPKH: info.address };
		case BtcAddressType.P2pkh:
			return { P2PKH: info.address };
		case BtcAddressType.P2sh:
			return { P2SH: info.address };
		case BtcAddressType.P2wsh:
			return { P2WSH: info.address };
		case BtcAddressType.P2tr:
			return { P2TR: info.address };
		default:
			// This should never happen if the BtcAddressType enum is exhaustive
			throw new Error(`Unsupported Bitcoin address type: ${info.type}`);
	}
};

export const parseBtcAddress = (address: string): BtcAddress | undefined => {
	try {
		const info = parseBtcAddressCkbtc({ address });
		return createBtcAddressFromAddressInfo({ info });
	} catch (_: unknown) {
		return;
	}
};

export const getBtcAddressString = (address: BtcAddress): string => {
	if ('P2WPKH' in address) {
		return address.P2WPKH;
	}
	if ('P2PKH' in address) {
		return address.P2PKH;
	}
	if ('P2SH' in address) {
		return address.P2SH;
	}
	if ('P2WSH' in address) {
		return address.P2WSH;
	}
	if ('P2TR' in address) {
		return address.P2TR;
	}

	return assertNever({ variable: address, typeName: 'BtcAddress' });
};
