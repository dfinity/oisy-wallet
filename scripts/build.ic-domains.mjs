import { readFileSync, writeFileSync } from 'node:fs';
import { join } from 'node:path';
import { OISY_IC_DOMAIN, replaceEnv } from './build.utils.mjs';

const generateDomain = (targetFile) => {
	let content = readFileSync(targetFile, 'utf8');

	const domain = OISY_IC_DOMAIN;

	// For information such as custom domains, only the domain without protocol is required
	const { host: value } = new URL(domain);
	content = replaceEnv({
		content,
		pattern: `{{VITE_OISY_DOMAIN}}`,
		value
	});

	writeFileSync(targetFile, content);
};

generateDomain(join(process.cwd(), 'build', '.well-known', 'ic-domains'));
