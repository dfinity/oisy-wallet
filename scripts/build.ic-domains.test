#!/usr/bin/env bash
set -euxo pipefail

# Prints all known dfx networks:
# - Built-in networks (local & ic)
# - Networks listed in dfx.json
dfx_networks() {
  echo ic
  echo local
  jq -r '.networks|keys[]' dfx.json
}

# Prints the Oisy domain name for every network.
# - The URL is the one in .well-known/ic-domains, used by the Internet Computer.
dfx_network_domains() {
  dfx_networks | while read -r i; do dfx build frontend --network "$i" >/dev/null 2>/dev/null && cat build/.well-known/ic-domains && echo " $i"; done
}

(
  echo Checking domains for all known networks...
  EXPECTED_DOMAINS="$0.expected"
  ACTUAL_DOMAINS="$(mktemp "$0.actual.XXXXXXXXXX")"
  dfx_network_domains >"$ACTUAL_DOMAINS"
  if diff --ignore-all-space --ignore-blank-lines <(sort <"$EXPECTED_DOMAINS") <(sort <"$ACTUAL_DOMAINS"); then
    echo OK: ic-domains
    rm "$ACTUAL_DOMAINS"
  else
    echo "ERROR: Incorrect .well-known/ic-domains"
    echo "Expected: $EXPECTED_DOMAINS"
    echo "Found:    $ACTUAL_DOMAINS"
    echo "==== START DIFF"
    diff "$EXPECTED_DOMAINS" "$ACTUAL_DOMAINS" || true
    echo "==== END DIFF"
    exit 1
  fi
)
