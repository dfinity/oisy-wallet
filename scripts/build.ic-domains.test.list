#!/usr/bin/env bash
set -euxo pipefail

# Prints all known dfx networks:
# - Built-in networks (local & ic)
# - Networks listed in dfx.json
echo ic
# echo local - Already listed in dfx.json
jq -r '.networks|keys[]' dfx.json
