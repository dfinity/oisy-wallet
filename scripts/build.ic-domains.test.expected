oisy.com ic
audit.oisy.com audit
beta.oisy.com beta
staging.oisy.com staging
staging.oisy.com test_be_1
fe1.oisy.com test_fe_1
fe2.oisy.com test_fe_2
fe3.oisy.com test_fe_3
fe4.oisy.com test_fe_4
e2e.oisy.com e2e
